include/eve/detail/function/simd/riscv/subscript.hpp

```diff
@@ -0,0 +1,112 @@
+//==================================================================================================
+/*
+  EVE - Expressive Vector Engine
+  Copyright : EVE Project Contributors
+  SPDX-License-Identifier: BSL-1.0
+*/
+//==================================================================================================
+#pragma once
+
+#include <eve/arch/riscv/rvv_common_masks.hpp>
+
+namespace eve::detail
+{
+template<callable_options O, typename T, typename N>
+EVE_FORCEINLINE T
+extract_(EVE_REQUIRES(rvv_), O const&, wide<T, N> v, std::size_t i) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  auto on_first_needed = __riscv_vslidedown_tu(v, v, i, N::value);
+  if constexpr( std::is_floating_point_v<T> ) return __riscv_vfmv_f(on_first_needed);
+  else return __riscv_vmv_x(on_first_needed);
+}
+
+template<callable_options O, typename T, typename N>
+EVE_FORCEINLINE void
+insert_(EVE_REQUIRES(rvv_),
+        O const&,
+        wide<T, N>                & v,
+        std::size_t                 i,
+        std::convertible_to<T> auto x) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  // get mask with 1 on i'th element.
+  logical<wide<T, N>> mask {[i](int j, int) { return j == i; }};
+  v = if_else(mask, static_cast<T>(x), v);
+}
+
+template<callable_options O, typename T, typename N>
+EVE_FORCEINLINE logical<T>
+extract_(EVE_REQUIRES(rvv_), O const&, logical<wide<T, N>> v, std::size_t i) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  constexpr size_t                      array_size = (N::value + 7) / 8;
+  std::array<unsigned char, array_size> mask_data;
+  __riscv_vsm(mask_data.data(), v, N::value);
+  size_t needed_element_id = i / 8;
+  return mask_data[needed_element_id] & (1 << (i % 8));
+}
+
+// For riscv logical we can not rely on common algorithm.
+template<callable_options O, typename T, typename N>
```

Comment:
This function should not be needed, default should do it. Why do you have it?

Reply:
I've touched this in https://github.com/jfalcou/eve/pull/1920#discussion_r1723584925 .

There is a problem in logicals. 

In common code we this:
https://github.com/jfalcou/eve/blob/318d0cbf1b0705c900173e6e4a0650d1bed2dbb4/include/eve/detail/function/simd/common/subscript.hpp#L36

where `abi_t::is_wide_logical` for RISC-V is `false`, so we fail into memcpy, that is wrong for RISC-V logical (as it is used `value_type`, not bool).

https://github.com/jfalcou/eve/blob/318d0cbf1b0705c900173e6e4a0650d1bed2dbb4/include/eve/detail/function/simd/common/subscript.hpp#L47

If I set `abi_t::is_wide_logical` to `true`, I will get errors in common slice funtionality:

https://github.com/jfalcou/eve/blob/318d0cbf1b0705c900173e6e4a0650d1bed2dbb4/include/eve/detail/function/simd/common/slice.hpp#L85

we have a wrong slice functions for logicals (as mask_type is wide), and bit_cast from `wide` to `logical` is not expected:

https://github.com/jfalcou/eve/blob/318d0cbf1b0705c900173e6e4a0650d1bed2dbb4/include/eve/detail/function/simd/common/slice.hpp#L89

Reply:
Never mind, It seems due to refactoring now this should work. Removed, thanks

Reply:
> Never mind, It seems due to refactoring now this should work. Removed, thanks

Finally fixed. The original problem was as I did not change 

https://github.com/jfalcou/eve/blob/318d0cbf1b0705c900173e6e4a0650d1bed2dbb4/include/eve/arch/cpu/tags.hpp#L55